Task 2/2.1/2.1.2/Polygons.cs

```diff
@@ -0,0 +1,139 @@
+        public Line(Point one, Point two) : base("Line", one, two) { }
+
+        public override double GetArea()
+        {
+            return 0;
+        }
+
+        public override string ToString()
+        {
+            return name + " - Point one: " + points[0] + ", two: " + points[1];
+        }
+
+        public bool isPerpendicular(Line argument)
+        {
+            double x = (this.points[0].x - this.points[1].x) * (argument.points[0].x - argument.points[1].x);
+            double y = (this.points[0].y - this.points[1].y) * (argument.points[0].y - argument.points[1].y);
+            if (x + y < 0.000001) return true;
+            return false;
+        }
+    }
+
+    public class Triangle : Polygon //треугольник
+    {
+        public Triangle(Point one, Point two, Point three) : base("Triangle", one, two, three) { }
+
+        public override double GetArea()
+        {
+            double p = GetLength()/2;
+            return Math.Sqrt(p*(p-points[0].GetDistance(points[1])) * (p - points[1].GetDistance(points[2])) * (p - points[2].GetDistance(points[0])));
+        }
+
+        public override string ToString()
+        {
+            string result = name + " - Point one: " + points[0] + ", two: " + points[1] + ", three: " + points[2];
+            return result + "; length: " + GetLength() + "; area: " + GetArea();
+        }
+    }
+
+    public abstract class Quadrilateral : Polygon //прямоугольное
+    {
+        public Quadrilateral(string name, Point one, Point two, Point three, Point four) : base(name, one, two, three, four) 
+        {
+            Line a = new Line(one, two);
+            Line b = new Line(two, three);
+            Line c = new Line(three, four);
+            Line d = new Line(four, one);
+            if (!a.isPerpendicular(b) || !b.isPerpendicular(c) || !c.isPerpendicular(d))
+            {
+                throw new Exception("The sides of " + name + " must be perpendicular");
+            }
+        }
+
+        public override double GetLength()
+        {
+            Line a = new Line(points[0], points[1]);
+            Line b = new Line(points[1], points[2]);
+            return 2*(a.GetLength() + b.GetLength());
+        }
+
+        public override double GetArea()
+        {
+            Line a = new Line(points[0], points[1]);
+            Line b = new Line(points[1], points[2]);
+            return a.GetLength() * b.GetLength();
+        }
+
+        public override string ToString()
+        {
+            string result = name + " - Point one: " + points[0] + ", two: " + points[1] + ", three: " + points[2] + ", four: " + points[3];
+            return result + "; length: " + GetLength() + "; area: " + GetArea();
+        }
+    }
+
+    public class Rectangle : Quadrilateral //прямоугольник
+    {
+        public Rectangle(Point one, Point two, Point three, Point four) : base("Rectangle", one, two, three, four) { }
+    }
+
+    public class Square : Quadrilateral //квадрат
+    {
```

Comment:
Same as with circles. Also I can replace one point after initialization (move it close to opposite) and square won't be square anymore, even though the type will be square - broken incapsulation - broken contract - GetArea returns wrong area because it hoped that the object is always square .
You could implement only Quad with additional Properties that say what kind of quad it is and universal area calculation algorithm. Or make it immutable.